flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -960,12 +965,26 @@ public void handleError(final Exception exception) {
 
 		@Override
 		public void releaseResource(InstanceID instanceId) {
-			stopWorker(instanceId);
+			runAsync(new Runnable() {
+				@Override
+				public void run() {
+					for (Map.Entry<ResourceID, WorkerRegistration<WorkerType>> entry : taskExecutors.entrySet()) {
+						if (entry.getValue().getInstanceID().equals(instanceId)) {
+							stopWorker(entry.getKey());
```

Comment:
I was thinking along those same lines - the slot manager deals with `InstanceID` mostly, and its log lines are tough to correlate with lower-level resource manager information that is `ResourceID` based.   It would be a nice improvement to make `InstanceID` a composite key that included `ResourceID`.

Regardless, I think the `stopWorker` method should use `ResourceID`, because `InstanceID` isn't a concept exposed to the RM subclasses.